### Evaluation of Search Algorithms
This notebook analyzes data generated from the five search algorithms used in the On The Town application. They are:
* Depth-First Search
* Breadth-First Search
* Greedy Search
* Uniform Cost Search
* A Star Search

The data were generated from the file `evaluation.ipynb`.

In [1]:
import pandas as pd
import plotly
import plotly.graph_objs as go
from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot
init_notebook_mode(connected=True)
plotly.tools.set_credentials_file(username='hangulu', api_key='xzmidZGzwl63Twl6sytL')

In [14]:
# Read the results from the CSV file
data = pd.read_csv("./data/simulation_results.csv")

In [15]:
data.head()

astar_sad  astar_time    bfs_sad   bfs_time    dfs_sad  dfs_time  \
0  17.280798    0.922846        NaN  60.002727  28.260684  0.261965   
1  10.489284    0.353055        NaN  60.066073        NaN  0.116575   
2  43.948272    0.037912  48.781254   0.151012        NaN  0.013806   
3  30.977607    0.118330        NaN  60.100918  54.002304  0.098764   
4  16.500775    0.397081        NaN  60.035823  22.037715  0.207314   

   greedy_sad  greedy_time  party_size    ucs_sad   ucs_time  
0   22.702997     0.254874           6  17.280798  56.156682  
1   11.851689     0.201409           6  10.489284  15.662877  
2   43.948272     0.039608           2  43.948272   0.171639  
3   32.120024     0.095724           3  30.977607   2.680794  
4   17.763021     0.231683           2  16.318736  48.145084

900 tests were run. Each test was limited to 60 seconds of evaluation time, and the test itself comprised of running all 5 algorithms, sequentially, and recording:
* whether a solution was reached
* the time each took to reach a solution
* the average sadness garnered by that solution. 

The following checks the percentage of times each algorithm returned no solution.

In [5]:
# DFS
dfs_no_soln = (data.dfs_sad.isna().sum()) / 900.
print "DFS failed " + str((dfs_no_soln * 100)) + "% of the time."

# BFS
bfs_no_soln = (data.bfs_sad.isna().sum()) / 900.
print "BFS failed " + str((bfs_no_soln * 100)) + "% of the time."

# Greedy Search
greedy_no_soln = (data.greedy_sad.isna().sum()) / 900.
print "Greedy Search failed " + str((greedy_no_soln * 100)) + "% of the time."

# UCS
ucs_no_soln = (data.ucs_sad.isna().sum()) / 900.
print "UCS failed " + str((ucs_no_soln * 100)) + "% of the time."

# A* Search
astar_no_soln = (data.astar_sad.isna().sum()) / 900.
print "A* Search failed " + str((astar_no_soln * 100)) + "% of the time."

DFS failed 32.0% of the time.
BFS failed 88.88888888888889% of the time.
Greedy Search failed 14.888888888888888% of the time.
UCS failed 39.666666666666664% of the time.
A* Search failed 15.88888888888889% of the time.


The following are the results:

* DFS failed 32.000% of the time.
* BFS failed 88.889% of the time.
* Greedy Search failed 14.889% of the time.
* UCS failed 39.667% of the time.
* A* Search failed 15.889% of the time.

Next, the average time elapsed at each simulation is analyzed. This analysis ignores the number of times the algorithms took more than 1 minute to complete, and thus calculates the average length of the completed solution space.

In [6]:
# DFS
avg_dfs_time = (data[data.dfs_time < 60].dfs_time.sum()) / (data[data.dfs_time < 60].dfs_time.count())
print "When completed, DFS took " + str(avg_dfs_time) + " seconds on average."

# BFS
avg_bfs_time = (data[data.bfs_time < 60].bfs_time.sum()) / (data[data.bfs_time < 60].bfs_time.count())
print "When completed, BFS took " + str(avg_bfs_time) + " seconds on average."

# Greedy
avg_greedy_time = (data[data.greedy_time < 60].greedy_time.sum()) / (data[data.greedy_time < 60].greedy_time.count())
print "When completed, Greedy Search took " + str(avg_greedy_time) + " seconds on average."

# UCS
avg_ucs_time = (data[data.ucs_time < 60].ucs_time.sum()) / (data[data.ucs_time < 60].ucs_time.count())
print "When completed, UCS took " + str(avg_ucs_time) + " seconds on average."

# A* Search
avg_astar_time = (data[data.astar_time < 60].astar_time.sum()) / (data[data.astar_time < 60].astar_time.count())
print "When completed, A* Search took " + str(avg_astar_time) + " seconds on average."

When completed, DFS took 0.28874309963650174 seconds on average.
When completed, BFS took 10.948256943560294 seconds on average.
When completed, Greedy Search took 1.047404620783899 seconds on average.
When completed, UCS took 15.10807837055115 seconds on average.
When completed, A* Search took 3.098107170896465 seconds on average.


The following are the results:

* When completed, DFS took 0.289 seconds on average.
* When completed, BFS took 10.948 seconds on average.
* When completed, Greedy Search took 1.047 seconds on average.
* When completed, UCS took 15.108 seconds on average.
* When completed, A* Search took 3.0981 seconds on average.

Next, the average sadness generated at each simulation is analyzed. Again, this analysis ignores the times the algorithms failed (took more than 1 minute or did not find a solution).

In [7]:
# DFS
avg_dfs_sad = data.dfs_sad.mean()
print "When completed, DFS produced an average sadness score of " + str(avg_dfs_sad) + "."

# BFS
avg_bfs_sad = data.bfs_sad.mean()
print "When completed, BFS produced an average sadness score of " + str(avg_bfs_sad) + "."

# Greedy
avg_greedy_sad = data.greedy_sad.mean()
print "When completed, Greedy Search produced an average sadness score of " + str(avg_greedy_sad) + "."

# UCS
avg_ucs_sad = data.ucs_sad.mean()
print "When completed, UCS produced an average sadness score of " + str(avg_ucs_sad) + "."

# A* Search
avg_astar_sad = data.astar_sad.mean()
print "When completed, A* Search produced an average sadness score of " + str(avg_astar_sad) + "."

When completed, DFS produced an average sadness score of 28.1894960099.
When completed, BFS produced an average sadness score of 26.9977561181.
When completed, Greedy Search produced an average sadness score of 22.5920907814.
When completed, UCS produced an average sadness score of 16.3321656619.
When completed, A* Search produced an average sadness score of 19.0470854323.


The following are the results:

* When completed, DFS produced an average sadness score of 28.189.
* When completed, BFS produced an average sadness score of 26.998.
* When completed, Greedy Search produced an average sadness score of 22.592.
* When completed, UCS produced an average sadness score of 16.332.
* When completed, A* Search produced an average sadness score of 19.047.

Next, the algorithms are compared graphically.

The following graphs are constructed for comparison:

* bar charts comparing the number of times each algorithm fails
* histograms showing the distribution of time each algorithm takes
* bar charts comparing the average time each algorithm takes
* histograms showing the distribution of sadness scores
* bar charts comparing the average sadness scores for each algorithm

In [8]:
# Bar chart for tests failed
trace = {
    'x': ['DFS', 'BFS', 'Greedy', 'UCS', 'A*'],  
    'y': [32.0, 88.88888888888889, 14.888888888888888, 39.666666666666664, 15.88888888888889],
    'type': 'bar',
    'marker': dict(color='rgb(255, 121, 121)')
}

chart = [trace]
layout = {
    'xaxis': {'title': 'Search Algorithm'},
    'yaxis': {'title': 'Percentage Of Tests Failed (%)'},
    'title': 'Percentage Failure Of Each Search Algorithm'
};
plotly.plotly.iplot({'data': chart, 'layout': layout}, filename='fail_bar')

In [9]:
# Histogram for time

dfs_time = data.dfs_time
bfs_time = data.bfs_time
greedy_time = data.greedy_time
ucs_time = data.ucs_time
astar_time = data.astar_time

trace0 = go.Histogram(x=dfs_time, name='DFS')
trace1 = go.Histogram(x=bfs_time, name='BFS')
trace2 = go.Histogram(x=greedy_time, name='Greedy')
trace3 = go.Histogram(x=ucs_time, name='UCS')
trace4 = go.Histogram(x=astar_time, name='A*')
  
fig = plotly.tools.make_subplots(rows=3, cols=2)
fig.append_trace(trace0, 1, 1)
fig.append_trace(trace1, 1, 2)
fig.append_trace(trace2, 2, 1)
fig.append_trace(trace3, 2, 2)
fig.append_trace(trace4, 3, 1)

fig['layout']['yaxis1'].update(title='Frequency', showgrid=False)
fig['layout']['yaxis2'].update(title='Frequency', showgrid=False)
fig['layout']['yaxis3'].update(title='Frequency', showgrid=False)
fig['layout']['yaxis4'].update(title='Frequency', showgrid=False)
fig['layout']['yaxis5'].update(title='Frequency', showgrid=False)

fig['layout']['xaxis1'].update(title='DFS Evaluation Time (seconds)', showgrid=False)
fig['layout']['xaxis2'].update(title='BFS Evaluation Time (seconds)', showgrid=False)
fig['layout']['xaxis3'].update(title='Greedy Evaluation Time (seconds)', showgrid=False)
fig['layout']['xaxis4'].update(title='UCS Evaluation Time (seconds)', showgrid=False)
fig['layout']['xaxis5'].update(title='A* Evaluation Time (seconds)', showgrid=False)
fig['layout'].update(title='Distribution Of Evaluation Time For Each Search Algorithm')

plotly.plotly.iplot(fig, filename='time_hist')

This is the format of your plot grid:
[ (1,1) x1,y1 ]  [ (1,2) x2,y2 ]
[ (2,1) x3,y3 ]  [ (2,2) x4,y4 ]
[ (3,1) x5,y5 ]  [ (3,2) x6,y6 ]



In [10]:
# Bar chart for time
trace = {
    'x': ['DFS', 'BFS', 'Greedy', 'UCS', 'A*'],  
    'y': [0.28874309963650174, 10.948256943560294, 1.047404620783899, 15.10807837055115, 3.098107170896465],
    'type': 'bar',
    'marker': dict(color='rgb(246, 229, 141)')
}

chart = [trace]
layout = {
    'xaxis': {'title': 'Search Algorithm'},
    'yaxis': {'title': 'Average Evaluation Time (seconds)'},
    'title': 'Average Evaluation Time Of Each Search Algorithm (When Complete)'
};
plotly.plotly.iplot({'data': chart, 'layout': layout}, filename='time_bar')

In [11]:
# Histogram for sadness

dfs_sad = data.dfs_sad
bfs_sad = data.bfs_sad
greedy_sad = data.greedy_sad
ucs_sad = data.ucs_sad
astar_sad = data.astar_sad

trace0 = go.Histogram(x=dfs_sad, name='DFS')
trace1 = go.Histogram(x=bfs_sad, name='BFS')
trace2 = go.Histogram(x=greedy_sad, name='Greedy')
trace3 = go.Histogram(x=ucs_sad, name='UCS')
trace4 = go.Histogram(x=astar_sad, name='A*')
  
fig = plotly.tools.make_subplots(rows=3, cols=2)
fig.append_trace(trace0, 1, 1)
fig.append_trace(trace1, 1, 2)
fig.append_trace(trace2, 2, 1)
fig.append_trace(trace3, 2, 2)
fig.append_trace(trace4, 3, 1)

fig['layout']['yaxis1'].update(title='Frequency', showgrid=False)
fig['layout']['yaxis2'].update(title='Frequency', showgrid=False)
fig['layout']['yaxis3'].update(title='Frequency', showgrid=False)
fig['layout']['yaxis4'].update(title='Frequency', showgrid=False)
fig['layout']['yaxis5'].update(title='Frequency', showgrid=False)

fig['layout']['xaxis1'].update(title='DFS Sadness', showgrid=False)
fig['layout']['xaxis2'].update(title='BFS Sadness', showgrid=False)
fig['layout']['xaxis3'].update(title='Greedy Sadness', showgrid=False)
fig['layout']['xaxis4'].update(title='UCS Sadness', showgrid=False)
fig['layout']['xaxis5'].update(title='A* Sadness', showgrid=False)
fig['layout'].update(title='Distribution Of Sadness For Each Search Algorithm')

plotly.plotly.iplot(fig, filename='sad_hist')

This is the format of your plot grid:
[ (1,1) x1,y1 ]  [ (1,2) x2,y2 ]
[ (2,1) x3,y3 ]  [ (2,2) x4,y4 ]
[ (3,1) x5,y5 ]  [ (3,2) x6,y6 ]



In [12]:
# Bar chart for sadness
trace = {
    'x': ['DFS', 'BFS', 'Greedy', 'UCS', 'A*'],  
    'y': [28.1894960099, 26.9977561181, 22.5920907814, 16.3321656619, 19.0470854323],
    'type': 'bar',
    'marker': dict(color='rgb(186, 220, 88)')
}

chart = [trace]
layout = {
    'xaxis': {'title': 'Search Algorithm'},
    'yaxis': {'title': 'Average Sadness'},
    'title': 'Average Sadness Of Each Search Algorithm (When Complete)'
};
plotly.plotly.iplot({'data': chart, 'layout': layout}, filename='sad_bar')

The results are discussed in detail in the **Results** section of the accompanying paper, but in short, Greedy Search is the best algorithm for this problem given its reliability (low percentage of failures), speed (second fastest), and effectiveness (third most satisfactory).